### PR TITLE
all client, support basic versioning for Operation, ModelProperty and Parameter

### PR DESCRIPTION
**Pull request by XiaofeiCao**
fix https://github.com/Azure/autorest.java/issues/2644

code in https://github.com/Azure/autorest.java/pull/2653/commits/01536598dde7316759619016dc3076c486d070cb

### 1. On Operation
```typescript
@removed(ApiVersions.v2022_12_01_preview)
create is ResourceOperations.ResourceCreateOrReplace<Resource>;
```

### 2. On ModelProperty
```typescript
@added(Versions.v2022_06_01_preview)
@removed(Versions.v2023_11_01)
@added(Versions.v2023_12_01_preview)
propertyRemovedInStable: string;
```

### 3. On Parameter
```typescript
model ListQueryParams {
  @removed(ApiVersions.v2022_12_01_preview)
  @query
  expand?: string;
}

op list is ResourceOperations.ResourceList<Resource, ListQueryParametersTrait<ListQueryParams>>;
```

### CODE REVIEW

File: typespec-tests/tsp/arm.tsp [hunk 1]
```diff
@@ -139,7 +177,16 @@ interface CustomTemplateResourceInterface
...
     > {
+  @sharedRoute
+  @added(Versions.v2022_09_01_preview)
+  @removed(Versions.v2023_11_01)
   update is ArmResourceUpdateSync<CustomTemplateResource>;
+
+  #suppress "@azure-tools/typespec-azure-core/no-operation-id" "Use operationId to keep same name as deprecated update operation"
+  @sharedRoute
+  @added(Versions.v2023_11_01)
+  @operationId("CustomTemplateResourceInterfaces_Update")
```
Comment (XiaofeiCao) [inline typespec-tests/tsp/arm.tsp:188]: example: https://github.com/Azure/azure-rest-api-specs/blob/d53388b8b7574dfe861a9e6863bd4d137168cbe5/specification/containerservice/Fleet.Management/fleetmember.tsp#L100

This is not supported yet. Will fix in following PR.

File: typespec-tests/tspconfig.yaml [hunk 2]
```diff
@@ -10,5 +10,4 @@ options:
     flavor: "Azure"
     dev-options:
       generate-code-model: true
-      support-versioning: true
```
Comment (weidongxu-microsoft) [inline typespec-tests/tspconfig.yaml:13]: I think you need to set this flag in script, for resiliency/servicedriven test -- they are for the advanced case to solve SDK API breaks (I am not sure whether tsp cmd can set the flag under a object, if not we can move this one level above).

File: typespec-extension/src/code-model-builder.ts [hunk 3]
```diff
@@ -667,7 +672,28 @@ export class CodeModelBuilder {
       this.trace(`Operation '${operation.name}' is temporary skipped, as it is an overloaded operation`);
       return true;
     }
-    return false;
+    return !this.existsAtCurrentVersion(operation) || false;
```
Comment (weidongxu-microsoft) [inline typespec-extension/src/code-model-builder.ts:675]: Seems easier to understand by putting it at top, e.g.

```
if (!this.existsAtCurrentVersion(operation)) {
  return true;
}
```

File: typespec-tests/tsp/arm.tsp (hunk 1 above)
Comment (weidongxu-microsoft) [reply typespec-tests/tsp/arm.tsp:188]: I don't know whether we should honor `@operationId`. Likely not, for greenfield.

File: typespec-tests/tspconfig.yaml (hunk 2 above)
Comment (XiaofeiCao) [reply typespec-tests/tspconfig.yaml:13]: Yes, it can. Updated.

File: typespec-extension/src/code-model-builder.ts (hunk 3 above)
Comment (XiaofeiCao) [reply typespec-extension/src/code-model-builder.ts]: fixed

File: typespec-tests/tspconfig.yaml (hunk 2 above)
Comment (weidongxu-microsoft) [reply typespec-tests/tspconfig.yaml:13]: It may be easier to disable it in tspconfig, and only enable on that particular test. Up to you.

Comment (XiaofeiCao) [reply typespec-tests/tspconfig.yaml:13]: OK. If we are using basic versioning for default mode, this is better.

Comment (weidongxu-microsoft) [reply typespec-tests/tspconfig.yaml:13]: Yeah, this is in dev-options, so it definitely not a default.

File: typespec-tests/Generate.ps1 [hunk 4]
```diff
@@ -38,11 +38,18 @@ $generateScript = {
   } elseif ($tspFile -match "resiliency[\\/]srv-driven[\\/]old.tsp") {
     # override namespace for "resiliency/srv-driven/old.tsp" (make it different to that from "main.tsp")
     $tspOptions += " --option ""@azure-tools/typespec-java.namespace=com.resiliency.servicedriven.v1"""
+    # enable advanced versioning for resiliency test
+    $tspOptions += " --option ""@azure-tools/typespec-java.dev-options.support-versioning=false"""
+  } elseif ($tspFile -match "resiliency[\\/]srv-driven[\\/]main.tsp") {
+    # enable advanced versioning for resiliency test
+    $tspOptions += " --option ""@azure-tools/typespec-java.dev-options.support-versioning=false"""
```
Comment (weidongxu-microsoft) [inline typespec-tests/Generate.ps1:45]: =true ?

Comment (XiaofeiCao) [reply typespec-tests/Generate.ps1]: My bad.. Fixed.

File: typespec-tests/tspconfig.yaml (hunk 2 above)
Comment (XiaofeiCao) [reply typespec-tests/tspconfig.yaml:13]: Emmm, seems I can't get it to work under an object... Maybe previously I got the test wrong(misconfigured tspconfig.yaml). Let me put it to the root level.

Comment (XiaofeiCao) [reply typespec-tests/tspconfig.yaml:13]: moved, renamed to `advanced-versioning` to distinguish between basic versioning.

File: typespec-extension/src/code-model-builder.ts [hunk 5]
```diff
@@ -2247,7 +2277,8 @@ export class CodeModelBuilder {
       if (
         prop.name === discriminatorPropertyName || // skip the discriminator property
         isNeverType(prop.type) || // skip property of type "never"
-        !isPayloadProperty(this.program, prop)
+        !isPayloadProperty(this.program, prop) ||
+        !this.existsAtCurrentVersion(prop)
```
Comment (weidongxu-microsoft) [inline typespec-extension/src/code-model-builder.ts:2281]: @haolingdong-msft for awareness. This may conflict with your Models.

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by haolingdong-msft** [comment]
double confirm my understanding: current support is to generate removed operations, which means even user switch to an old service version which contains a removed operation, they can still not call the newly removed operation?

**Comment by weidongxu-microsoft** [comment]
Yes.

For mgmt, this is same behavior of current Swagger, and there is no "switch to an old service version".

For data-plane, we don't plan to address the "remove" in near future. Service should not "remove" anything that GAed.
For the preview api-versions before API, current guideline is to delete them in typespec (as preview api-version should only live for about 3 months in data-plane).

**Pull request merged**